File: cliffy-protocols/src/lattice.rs
//! Lattice-based conflict resolution using geometric algebra
//!
//! This module provides the `GeometricLattice` trait for join-semilattice operations
//! that enable coordination-free conflict resolution in distributed systems.
//!
//! # Key Properties
//!
//! A join-semilattice must satisfy:
//! - **Idempotent**: `a ⊔ a = a`
//! - **Commutative**: `a ⊔ b = b ⊔ a`
//! - **Associative**: `(a ⊔ b) ⊔ c = a ⊔ (b ⊔ c)`
//!
//! These properties ensure that replicas always converge regardless of message
//! ordering or network partitions.
//!
//! # Example
//!
//! ```rust
//! use cliffy_protocols::lattice::{GeometricLattice, GA3Lattice};
//! use cliffy_core::GA3;
//!
//! // Create two conflicting states
//! let state_a = GA3Lattice::from_scalar(1.0);
//! let state_b = GA3Lattice::from_scalar(2.0);
//!
//! // Join always produces a consistent result
//! let joined = state_a.join(&state_b);
//! assert!(joined.dominates(&state_a));
//! assert!(joined.dominates(&state_b));
//! ```

use crate::geometric_mean;
use cliffy_core::GA3;

/// A join-semilattice with geometric algebra operations.
///
/// This trait provides the mathematical foundation for CRDTs:
/// - `join` computes the least upper bound (always converges)
/// - `dominates` checks causal ordering
/// - `divergence` measures conflict severity
pub trait GeometricLattice: Clone {
    /// Lattice join (least upper bound) - always converges, no coordination needed.
    ///
    /// The join operation must be idempotent, commutative, and associative.
    fn join(&self, other: &Self) -> Self;

    /// Check if this state dominates (is greater than or equal to) another.
    ///
    /// Returns true if `other ⊔ self = self`.
    fn dominates(&self, other: &Self) -> bool;

    /// Compute the geometric distance/divergence from another state.
    ///
    /// This measures how "far apart" two states are, useful for:
    /// - Detecting conflicts
    /// - Measuring convergence progress
    /// - Prioritizing sync operations
    fn divergence(&self, other: &Self) -> f64;

    /// Check if two states are equal in the lattice ordering.
    fn lattice_eq(&self, other: &Self) -> bool {
        self.dominates(other) && other.dominates(self)
    }

    /// Compute the lattice meet (greatest lower bound) if it exists.
    ///
    /// Not all semilattices have meets, so this returns Option.
    fn meet(&self, other: &Self) -> Option<Self>;
}

/// A wrapper around GA3 that implements GeometricLattice.
///
/// This provides lattice operations for multivectors where:
/// - Join uses geometric mean for equal-magnitude states
/// - Dominance is based on magnitude ordering
/// - Divergence is the geometric distance
#[derive(Debug, Clone, PartialEq)]
pub struct GA3Lattice {
    inner: GA3,
}

impl GA3Lattice {
    /// Create a new lattice element from a multivector.
    pub fn new(mv: GA3) -> Self {
        Self { inner: mv }
    }

    /// Create a lattice element from a scalar.
    pub fn from_scalar(value: f64) -> Self {
        Self::new(GA3::scalar(value))
    }

    /// Create a lattice element from vector components.
    pub fn from_vector(x: f64, y: f64, z: f64) -> Self {
        use amari_core::Vector;
        let v = Vector::<3, 0, 0>::from_components(x, y, z);
        Self::new(GA3::from_vector(&v))
    }

    /// Create the zero element (bottom of the lattice).
    pub fn zero() -> Self {
        Self::new(GA3::zero())
    }

    /// Get the underlying multivector.
    pub fn as_multivector(&self) -> &GA3 {
        &self.inner
    }

    /// Consume and return the underlying multivector.
    pub fn into_multivector(self) -> GA3 {
        self.inner
    }

    /// Get the magnitude of this lattice element.
    pub fn magnitude(&self) -> f64 {
        self.inner.magnitude()
    }

    /// Get a coefficient at the given index.
    pub fn get(&self, index: usize) -> f64 {
        self.inner.get(index)
    }
}

impl GeometricLattice for GA3Lattice {
    fn join(&self, other: &Self) -> Self {
        // Check for structural equality first (idempotence optimization)
        if self.divergence(other) < 1e-10 {
            return self.clone();
        }

        let self_mag = self.inner.magnitude();
        let other_mag = other.inner.magnitude();

        // Dominance by magnitude
        if self_mag > other_mag + 1e-10 {
            self.clone()
        } else if other_mag > self_mag + 1e-10 {
            other.clone()
        } else {
            // Equal magnitudes but different states - use geometric mean
            Self::new(geometric_mean(&[self.inner.clone(), other.inner.clone()]))
        }
    }

    fn dominates(&self, other: &Self) -> bool {
        // A dominates B if magnitude(A) >= magnitude(B)
        // This creates a total order for non-zero states
        self.inner.magnitude() >= other.inner.magnitude() - 1e-10
    }

    fn divergence(&self, other: &Self) -> f64 {
        (&self.inner - &other.inner).magnitude()
    }

    fn meet(&self, other: &Self) -> Option<Self> {
        let self_mag = self.inner.magnitude();
        let other_mag = other.inner.magnitude();

        // Meet is the element with smaller magnitude
        if self_mag < other_mag + 1e-10 {
            Some(self.clone())
        } else if other_mag < self_mag + 1e-10 {
            Some(other.clone())
        } else {
            // Equal magnitudes - meet exists and equals both
            Some(self.clone())
        }
    }
}

/// Component-wise lattice operations for multivectors.
///
/// Unlike `GA3Lattice` which uses magnitude ordering, this provides
/// coefficient-by-coefficient join/meet operations.
#[derive(Debug, Clone, PartialEq)]
pub struct ComponentLattice {
    inner: GA3,
}

impl ComponentLattice {
    /// Create a new component lattice element.
    pub fn new(mv: GA3) -> Self {
        Self { inner: mv }
    }

    /// Create from a scalar value.
    pub fn from_scalar(value: f64) -> Self {
        Self::new(GA3::scalar(value))
    }

    /// Get the underlying multivector.
    pub fn as_multivector(&self) -> &GA3 {
        &self.inner
    }

    /// Consume and return the underlying multivector.
    pub fn into_multivector(self) -> GA3 {
        self.inner
    }
}

impl GeometricLattice for ComponentLattice {
    fn join(&self, other: &Self) -> Self {
        // Component-wise maximum
        let mut coeffs = Vec::with_capacity(8);
        for i in 0..8 {
            coeffs.push(self.inner.get(i).max(other.inner.get(i)));
        }
        Self::new(GA3::from_coefficients(coeffs))
    }

    fn dominates(&self, other: &Self) -> bool {
        // Dominates if every component is >= the corresponding component
        (0..8).all(|i| self.inner.get(i) >= other.inner.get(i) - 1e-10)
    }

    fn divergence(&self, other: &Self) -> f64 {
        // L-infinity norm (max component difference)
        (0..8)
            .map(|i| (self.inner.get(i) - other.inner.get(i)).abs())
            .fold(0.0_f64, |acc, d| acc.max(d))
    }

    fn meet(&self, other: &Self) -> Option<Self> {
        // Component-wise minimum
        let mut coeffs = Vec::with_capacity(8);
        for i in 0..8 {
            coeffs.push(self.inner.get(i).min(other.inner.get(i)));
        }
        Some(Self::new(GA3::from_coefficients(coeffs)))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_ga3_lattice_idempotent() {
        let a = GA3Lattice::from_scalar(5.0);
        let joined = a.join(&a);
        assert!(a.lattice_eq(&joined));
    }

    #[test]
    fn test_ga3_lattice_commutative() {
        let a = GA3Lattice::from_scalar(3.0);
        let b = GA3Lattice::from_scalar(7.0);

        let ab = a.join(&b);
        let ba = b.join(&a);

        assert!(ab.lattice_eq(&ba));
    }

    #[test]
    fn test_ga3_lattice_associative() {
        let a = GA3Lattice::from_scalar(2.0);
        let b = GA3Lattice::from_scalar(5.0);
        let c = GA3Lattice::from_scalar(3.0);

        let ab_c = a.join(&b).join(&c);
        let a_bc = a.join(&b.join(&c));

        assert!(ab_c.lattice_eq(&a_bc));
    }

    #[test]
    fn test_ga3_lattice_dominance() {
        let small = GA3Lattice::from_scalar(2.0);
        let large = GA3Lattice::from_scalar(5.0);

        assert!(large.dominates(&small));
        assert!(!small.dominates(&large));
    }

    #[test]
    fn test_ga3_lattice_divergence() {
        let a = GA3Lattice::from_scalar(3.0);
        let b = GA3Lattice::from_scalar(7.0);

        let div = a.divergence(&b);
        assert!((div - 4.0).abs() < 1e-10);
    }

    #[test]
    fn test_component_lattice_idempotent() {
        let a = ComponentLattice::from_scalar(5.0);
        let joined = a.join(&a);
        assert!(a.lattice_eq(&joined));
    }

    #[test]
    fn test_component_lattice_commutative() {
        let a = ComponentLattice::new(GA3::from_coefficients(vec![
            1.0, 2.0, 3.0, 0.0, 0.0, 0.0, 0.0, 0.0,
        ]));
        let b = ComponentLattice::new(GA3::from_coefficients(vec![
            2.0, 1.0, 4.0, 0.0, 0.0, 0.0, 0.0, 0.0,
        ]));

        let ab = a.join(&b);
        let ba = b.join(&a);

        assert!(ab.lattice_eq(&ba));
    }

    #[test]
    fn test_component_lattice_join_max() {
        let a = ComponentLattice::new(GA3::from_coefficients(vec![
            1.0, 2.0, 3.0, 0.0, 0.0, 0.0, 0.0, 0.0,
        ]));
        let b = ComponentLattice::new(GA3::from_coefficients(vec![
            2.0, 1.0, 4.0, 0.0, 0.0, 0.0, 0.0, 0.0,
        ]));

        let joined = a.join(&b);

        assert!((joined.as_multivector().get(0) - 2.0).abs() < 1e-10);
        assert!((joined.as_multivector().get(1) - 2.0).abs() < 1e-10);
        assert!((joined.as_multivector().get(2) - 4.0).abs() < 1e-10);
    }

    #[test]
    fn test_component_lattice_meet_min() {
        let a = ComponentLattice::new(GA3::from_coefficients(vec![
            1.0, 2.0, 3.0, 0.0, 0.0, 0.0, 0.0, 0.0,
        ]));
        let b = ComponentLattice::new(GA3::from_coefficients(vec![
            2.0, 1.0, 4.0, 0.0, 0.0, 0.0, 0.0, 0.0,
        ]));

        let met = a.meet(&b).unwrap();

        assert!((met.as_multivector().get(0) - 1.0).abs() < 1e-10);
        assert!((met.as_multivector().get(1) - 1.0).abs() < 1e-10);
        assert!((met.as_multivector().get(2) - 3.0).abs() < 1e-10);
    }

    #[test]
    fn test_lattice_convergence() {
        // Simulate distributed updates converging
        let initial = GA3Lattice::from_scalar(1.0);

        // Two nodes make concurrent updates
        let node1_update = GA3Lattice::from_scalar(5.0);
        let node2_update = GA3Lattice::from_scalar(3.0);

        // Both nodes join with initial
        let node1_state = initial.join(&node1_update);
        let node2_state = initial.join(&node2_update);

        // Cross-sync: both should converge to the same state
        let node1_final = node1_state.join(&node2_state);
        let node2_final = node2_state.join(&node1_state);

        assert!(node1_final.lattice_eq(&node2_final));
    }
}

File: cliffy-tsukoshi/src/protocols/lattice.ts
/**
 * Lattice-based conflict resolution using geometric algebra.
 *
 * Join-semilattice operations enable coordination-free conflict resolution
 * in distributed systems with guaranteed convergence.
 *
 * Key Properties:
 * - Idempotent: a ⊔ a = a
 * - Commutative: a ⊔ b = b ⊔ a
 * - Associative: (a ⊔ b) ⊔ c = a ⊔ (b ⊔ c)
 *
 * @example
 * ```typescript
 * import { GA3Lattice } from 'cliffy-tsukoshi/protocols';
 *
 * const stateA = GA3Lattice.fromScalar(1);
 * const stateB = GA3Lattice.fromScalar(2);
 *
 * // Join always produces consistent result
 * const joined = stateA.join(stateB);
 * console.log(joined.dominates(stateA)); // true
 * console.log(joined.dominates(stateB)); // true
 * ```
 */

import type { GA3 } from '../ga3.js';
import {
  zero,
  scalar,
  vector,
  fromCoefficients,
  magnitude,
  sub,
  clone,
} from '../ga3.js';
import { geometricMean } from './crdt.js';

const EPSILON = 1e-10;

/**
 * Interface for geometric lattice operations.
 */
export interface GeometricLattice<T> {
  /** Lattice join (least upper bound) - always converges */
  join(other: T): T;

  /** Check if this state dominates (is >= to) another */
  dominates(other: T): boolean;

  /** Compute the geometric distance/divergence from another state */
  divergence(other: T): number;

  /** Check if two states are equal in the lattice ordering */
  latticeEq(other: T): boolean;

  /** Compute the lattice meet (greatest lower bound) if it exists */
  meet(other: T): T | null;
}

/**
 * A wrapper around GA3 that implements GeometricLattice.
 *
 * Provides lattice operations where:
 * - Join uses geometric mean for equal-magnitude states
 * - Dominance is based on magnitude ordering
 * - Divergence is the geometric distance
 */
export class GA3Lattice implements GeometricLattice<GA3Lattice> {
  private inner: GA3;

  constructor(mv: GA3) {
    this.inner = mv;
  }

  /** Create a lattice element from a scalar. */
  static fromScalar(value: number): GA3Lattice {
    return new GA3Lattice(scalar(value));
  }

  /** Create a lattice element from vector components. */
  static fromVector(x: number, y: number, z: number): GA3Lattice {
    return new GA3Lattice(vector(x, y, z));
  }

  /** Create the zero element (bottom of the lattice). */
  static zero(): GA3Lattice {
    return new GA3Lattice(zero());
  }

  /** Get the underlying multivector. */
  asMultivector(): GA3 {
    return this.inner;
  }

  /** Get the magnitude of this lattice element. */
  magnitude(): number {
    return magnitude(this.inner);
  }

  /** Get a coefficient at the given index. */
  get(index: number): number {
    return (this.inner as number[])[index];
  }

  /** Create a copy of this lattice element. */
  clone(): GA3Lattice {
    return new GA3Lattice(clone(this.inner));
  }

  /** Lattice join (least upper bound). */
  join(other: GA3Lattice): GA3Lattice {
    // Check for structural equality first (idempotence optimization)
    if (this.divergence(other) < EPSILON) {
      return this.clone();
    }

    const selfMag = magnitude(this.inner);
    const otherMag = magnitude(other.inner);

    // Dominance by magnitude
    if (selfMag > otherMag + EPSILON) {
      return this.clone();
    } else if (otherMag > selfMag + EPSILON) {
      return other.clone();
    } else {
      // Equal magnitudes but different states - use geometric mean
      return new GA3Lattice(geometricMean([this.inner, other.inner]));
    }
  }

  /** Check if this state dominates another. */
  dominates(other: GA3Lattice): boolean {
    return magnitude(this.inner) >= magnitude(other.inner) - EPSILON;
  }

  /** Compute the divergence (geometric distance) from another state. */
  divergence(other: GA3Lattice): number {
    return magnitude(sub(this.inner, other.inner));
  }

  /** Check if two states are equal in the lattice ordering. */
  latticeEq(other: GA3Lattice): boolean {
    return this.dominates(other) && other.dominates(this);
  }

  /** Compute the lattice meet (greatest lower bound). */
  meet(other: GA3Lattice): GA3Lattice | null {
    const selfMag = magnitude(this.inner);
    const otherMag = magnitude(other.inner);

    // Meet is the element with smaller magnitude
    if (selfMag < otherMag + EPSILON) {
      return this.clone();
    } else if (otherMag < selfMag + EPSILON) {
      return other.clone();
    } else {
      // Equal magnitudes - meet exists and equals both
      return this.clone();
    }
  }

  /** Serialize to JSON. */
  toJSON(): number[] {
    return this.inner as number[];
  }

  /** Deserialize from JSON. */
  static fromJSON(data: number[]): GA3Lattice {
    return new GA3Lattice(fromCoefficients(data));
  }
}

/**
 * Component-wise lattice operations for multivectors.
 *
 * Unlike GA3Lattice which uses magnitude ordering, this provides
 * coefficient-by-coefficient join/meet operations.
 */
export class ComponentLattice implements GeometricLattice<ComponentLattice> {
  private inner: GA3;

  constructor(mv: GA3) {
    this.inner = mv;
  }

  /** Create from a scalar value. */
  static fromScalar(value: number): ComponentLattice {
    return new ComponentLattice(scalar(value));
  }

  /** Get the underlying multivector. */
  asMultivector(): GA3 {
    return this.inner;
  }

  /** Create a copy. */
  clone(): ComponentLattice {
    return new ComponentLattice(clone(this.inner));
  }

  /** Component-wise maximum (join). */
  join(other: ComponentLattice): ComponentLattice {
    const selfCoeffs = this.inner as number[];
    const otherCoeffs = other.inner as number[];
    const result = selfCoeffs.map((c, i) => Math.max(c, otherCoeffs[i]));
    return new ComponentLattice(fromCoefficients(result));
  }

  /** Dominates if every component is >= the corresponding component. */
  dominates(other: ComponentLattice): boolean {
    const selfCoeffs = this.inner as number[];
    const otherCoeffs = other.inner as number[];
    return selfCoeffs.every((c, i) => c >= otherCoeffs[i] - EPSILON);
  }

  /** L-infinity norm (max component difference). */
  divergence(other: ComponentLattice): number {
    const selfCoeffs = this.inner as number[];
    const otherCoeffs = other.inner as number[];
    return Math.max(...selfCoeffs.map((c, i) => Math.abs(c - otherCoeffs[i])));
  }

  /** Check if two states are equal in the lattice ordering. */
  latticeEq(other: ComponentLattice): boolean {
    return this.dominates(other) && other.dominates(this);
  }

  /** Component-wise minimum (meet). */
  meet(other: ComponentLattice): ComponentLattice {
    const selfCoeffs = this.inner as number[];
    const otherCoeffs = other.inner as number[];
    const result = selfCoeffs.map((c, i) => Math.min(c, otherCoeffs[i]));
    return new ComponentLattice(fromCoefficients(result));
  }

  /** Serialize to JSON. */
  toJSON(): number[] {
    return this.inner as number[];
  }

  /** Deserialize from JSON. */
  static fromJSON(data: number[]): ComponentLattice {
    return new ComponentLattice(fromCoefficients(data));
  }
}

/**
 * Compute component-wise lattice join of two multivectors.
 */
export function latticeJoin(a: GA3, b: GA3): GA3 {
  const aCoeffs = a as number[];
  const bCoeffs = b as number[];
  return fromCoefficients(aCoeffs.map((c, i) => Math.max(c, bCoeffs[i])));
}

/**
 * Compute component-wise lattice meet of two multivectors.
 */
export function latticeMeet(a: GA3, b: GA3): GA3 {
  const aCoeffs = a as number[];
  const bCoeffs = b as number[];
  return fromCoefficients(aCoeffs.map((c, i) => Math.min(c, bCoeffs[i])));
}
